pkg/solana/txm/pendingtx.go

```diff
@@ -3,78 +3,122 @@ package txm
+	state       TxState
 }
 
 var _ PendingTxContext = &pendingTxContext{}
 
 type pendingTxContext struct {
-	cancelBy  map[uuid.UUID]context.CancelFunc
-	timestamp map[uuid.UUID]time.Time
-	sigToID   map[solana.Signature]uuid.UUID
-	idToSigs  map[uuid.UUID][]solana.Signature
-	lock      sync.RWMutex
+	cancelBy map[string]context.CancelFunc
+	sigToID  map[solana.Signature]string
+
+	broadcastedTxs      map[string]pendingTx // transactions that require retry and bumping i.e broadcasted, processed
+	confirmedTxs        map[string]pendingTx // transactions that require monitoring for re-org
+	finalizedErroredTxs map[string]pendingTx // finalized and errored transactions held onto for status
+
+	lock sync.RWMutex
 }
 
 func newPendingTxContext() *pendingTxContext {
 	return &pendingTxContext{
-		cancelBy:  map[uuid.UUID]context.CancelFunc{},
-		timestamp: map[uuid.UUID]time.Time{},
-		sigToID:   map[solana.Signature]uuid.UUID{},
-		idToSigs:  map[uuid.UUID][]solana.Signature{},
+		cancelBy: map[string]context.CancelFunc{},
+		sigToID:  map[solana.Signature]string{},
+
+		broadcastedTxs:      map[string]pendingTx{},
+		confirmedTxs:        map[string]pendingTx{},
+		finalizedErroredTxs: map[string]pendingTx{},
 	}
 }
 
-func (c *pendingTxContext) New(sig solana.Signature, cancel context.CancelFunc) (uuid.UUID, error) {
-	// validate signature does not exist
+func (c *pendingTxContext) New(tx pendingTx, sig solana.Signature, cancel context.CancelFunc) error {
 	c.lock.RLock()
+	// validate signature does not exist
 	if _, exists := c.sigToID[sig]; exists {
 		c.lock.RUnlock()
-		return uuid.UUID{}, errors.New("signature already exists")
+		return errors.New("signature already exists")
+	}
+	// validate id does not exist
+	if _, exists := c.broadcastedTxs[tx.id]; exists {
+		c.lock.RUnlock()
+		return errors.New("tx id already exists in broadcasted map")
 	}
 	c.lock.RUnlock()
 
-	// upgrade to write lock if sig does not exist
+	// upgrade to write lock if sig or id do not exist
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	if _, exists := c.sigToID[sig]; exists {
-		return uuid.UUID{}, errors.New("signature already exists")
+		return errors.New("signature already exists")
+	}
+	if _, exists := c.broadcastedTxs[tx.id]; exists {
+		return errors.New("tx id already exists in broadcasted map")
 	}
 	// save cancel func
-	id := uuid.New()
-	c.cancelBy[id] = cancel
-	c.timestamp[id] = time.Now()
-	c.sigToID[sig] = id
-	c.idToSigs[id] = []solana.Signature{sig}
-	return id, nil
+	c.cancelBy[tx.id] = cancel
+	c.sigToID[sig] = tx.id
+	// add signature to tx
+	tx.signatures = append(tx.signatures, sig)
+	tx.createTs = time.Now()
+	tx.state = Broadcasted
+	// save to the broadcasted map since transaction was just broadcasted
+	c.broadcastedTxs[tx.id] = tx
+	return nil
 }
 
-func (c *pendingTxContext) Add(id uuid.UUID, sig solana.Signature) error {
-	// already exists
+func (c *pendingTxContext) AddSignature(id string, sig solana.Signature) error {
 	c.lock.RLock()
```

Comment:
nit: would make sense to encapsulate locking logic into functions? It may reduce a couple of lines and make the locked portion more explicit: 

Something like: 

```
func (c *pendingTxContext) withWriteLock(fn func()) {
    c.lock.Lock()
    defer c.lock.Unlock()
    fn()
}

func (c *pendingTxContext) withReadLock(fn func()) {
    c.lock.RLock()
    defer c.lock.RUnlock()
    fn()
}

```
that could be used as:

```
c.withReadLock(func() {
    // Perform read operations
})

c.withWriteLock(func() {
    // Perform write operations
})
```

Reply:
I like this suggestion. I updated the code to use this in the latest commit.